ext/reflection/php_reflection.c

```diff
@@ -1956,7 +1958,7 @@ ZEND_METHOD(ReflectionFunctionAbstract, getAttributes)
 
 	reflect_attributes(INTERNAL_FUNCTION_PARAM_PASSTHRU,
 		fptr->common.attributes, 0, fptr->common.scope, target,
-		fptr->type == ZEND_USER_FUNCTION ? fptr->op_array.filename : NULL);
+		fptr->type == ZEND_USER_FUNCTION ? fptr->op_array.filename : NULL, ZEND_USER_CODE(fptr->type));
```

Comment:
Actually, I'm a bit confused by the intention here. Does this check whether the attribute class is internal, or whether the target is internal? It currently does the latter, but that doesn't seem super useful?

Reply:
The intention is more : does the attribute as been defined in user land code,

If it's an attribute from user land on an user land code it should return true
If it's an attribute from php core on an user land code it should return true
If it's an attribute from php core on php core it should return false
If it's an attribute from user land code on php core it should return false (but i don't think this use case is possible)

Reply:
Can you clarify why this matters? I understand that internal attributes can have additional validation without instanciation. But why does it matter whether the attrbute is placed on a user or internal target?

Reply:
This is mainly if some library that do error handling (like https://github.com/symfony/error-handler by example) want to provide a generic way to pin point those error, they could then only do that when they detect that the attribute has been writed by the user and 

But i'm wondering if PHP internal class may have internal attributes attached to their method or class (now or in the future) defined ? I know there is pre defined attributes by PHP but not sure if there are used in internal methods or if an extension can declare them using C code ?

If not then this is useless

Reply:
Yes, internal attributes can be attached to internal classes, but instantiating them should never error. If it does error, that would be a bug in core/the extension itself. I'm still confused about when and why exactly Symfony error handler would need this information.

Can you provide a concrete example? Moreover, whether the target is internal would already be known, by looking at the reflection of the class/method itself.

Reply:
As an example we could, in a generic error handler, do the following interface : 

```php
interface LinkReflectionAttribute
{
   public function getAttribute(): \ReflectionAttribute;
}
```

When the error handler catch an exception with this interface it may do special things like show the exception (like before) and add a special text (or html in web context) to the user to show which attributes this exception concern (it could even show the portion of code in web context).

Library would only have to implement this interface in their exception and the error handler would do the rest

Having this method would allow this generic handler to be sure that the reflection attribute given has been written in a php file and does not originate from the core (in the last case there is no point showing something since it's not written in a php file and so it's not fixable by the user)

This method may not be necessary if the interface look like this : 

```php
interface LinkReflectionAttribute
{
   public function getAttribute(): \ReflectionAttribute;
   
   public function getTarget(): \ReflectionClass | \ReflectionMethod |\ReflectionFunction | \ReflectionParameter | \ReflectionProperty
}
```

In this case we could determin if it's user defined given the specific target.

But maybe the `isUserDefined()` method is not the best approach. Wouldn't it be better if instead we have this `getTarget()` method inside the `ReflectionAttribute` class ?

So we could have the target of the attribute without having to keep track of it in php code ?

There is already a `getTarget()` method that returns an `int`, there could be a `getDeclaringTarget()` that returns `\ReflectionClass | \ReflectionMethod |\ReflectionFunction | \ReflectionParameter | \ReflectionProperty`